volume-migration: avoid to start a vm on failed migration

alicefr opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
Possible to start a VM after a failed volume migration with an inconsistent volume set. In this case, the VM could be started from an unbootable disk

After this PR:
Avoid to start a VM if the ManualRecoveryRequiered is set.


### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

alicefr commented:
/hold
It requires https://github.com/kubevirt/kubevirt/pull/12355 to be merged

alicefr commented:
/uncc @0xFelix 
/cc @mhenriks @vladikr

fossedihelm commented:
@alicefr Are you able to rebase this on top of the previous one?
As @vladikr pointed out, we should have this in the same release. Thanks

alicefr commented:
/retest

vladikr commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -472,6 +472,13 @@ func (app *SubresourceAPIApp) StartVMRequestHandler(request *restful.Request, re
+		vm.Status.VolumeUpdateState.VolumeMigrationState != nil &&
+		vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired != nil &&
+		*vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired {
+		writeError(errors.NewConflict(v1.Resource("virtualmachine"), name, fmt.Errorf("VM needs to be manually recovered because the volume migration failed and the volumes are in an incosistent state")), response)
```
maybe something like "VM recovery required: Volume migration failed, leaving some volumes pointing to non-existent targets; manual intervention is needed to reassign them to their original volumes." ?

vladikr commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -1621,10 +1621,32 @@ func syncVolumeMigration(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineIn
-	if vmi == nil && vmCond.HasConditionWithStatus(vm, virtv1.VirtualMachineConditionType(virtv1.VirtualMachineInstanceVolumesChange), k8score.ConditionTrue) {
+
+	srcMigVols := make(map[string]string)
+	for _, v := range vm.Status.VolumeUpdateState.VolumeMigrationState.MigratedVolumes {
```
maybe it needs a simple comment, "// Checks if VM volumes have been recovered to their original PVCs after a failed migration. "

vladikr reviewed (commented):
The PR looks solid to me. Thank @alicefr.
Just a few nits. 

It begs a question though - if we already know the source PVCs, can't we recover the volumes automatically?

fossedihelm commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -1621,10 +1621,32 @@ func syncVolumeMigration(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineIn
+	if recoveredOldVMVolumes {
+		vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired = nil
+		return
+	} else if vmi == nil && vmCond.HasConditionWithStatus(vm, virtv1.VirtualMachineConditionType(virtv1.VirtualMachineInstanceVolumesChange), k8score.ConditionTrue) {
```
nit: the else can be dropped
```suggestion
	}

	if vmi == nil && vmCond.HasConditionWithStatus(vm, virtv1.VirtualMachineConditionType(virtv1.VirtualMachineInstanceVolumesChange), k8score.ConditionTrue) {
		// Something went wrong with the VMI while the volume migration was in progress
		vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired = pointer.P(true)
	}
```

fossedihelm reviewed (commented):
Thank you @alicefr!
Looks good overall. Apart from the small nit below, I have two things:
1. Shouldn't we consider the restart too? I mean, if a VM is restarted during the migration, shouldn't we remain in an inconsistent state?
2. Maybe we are missing a unit test in `subresource_test`

Thanks

alicefr commented:
> It begs a question though - if we already know the source PVCs, can't we recover the volumes automatically?

No, because it will be a violation of the declarative approach we have been following for this feature. In order to recover the volumes automatically, we would need to patch the volume spec in behalf of the users.

alicefr commented:
> 2. Maybe we are missing a unit test in `subresource_test`
@fossedihelm Done, PTAL

vladikr commented:
> > It begs a question though - if we already know the source PVCs, can't we recover the volumes automatically?
> 
> No, because it will be a violation of the declarative approach we have been following for this feature. In order to recover the volumes automatically, we would need to patch the volume in the VM spec in behalf of the users.
> 
> EDIT: this should be done by the overlaying tools for the migration plans

I meant that we would boot from the known source volume from the status and then the migration would be re-triggered.
This wouldn't violate the declarative approach but would improve the user experience.

However, I think this can be done in the future.
Currently, I'm more concerned about VM potentially reaching an unbootable state.

/approve

mhenriks commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -1621,10 +1621,34 @@ func syncVolumeMigration(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineIn
+		}
+	}
+	if recoveredOldVMVolumes {
+		vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired = nil
```
Unit test for this?

alicefr replied:
I need to add entirely for the sycVolumeMigration function? Can we do it as a follow back?

mhenriks replied:
I don't see any coverage in unittests for `status.volumeUpdateState` but can be done in follow up

alicefr replied:
I added the unit tests please, check the last commit

alicefr commented:
@fossedihelm the issue is that the test isn't added by this PR but it is already upstream. Then, we have a flake :/

alicefr commented:
/retest pull-kubevirt-e2e-k8s-1.31-sig-storage

fossedihelm commented on pkg/virt-api/rest/subresource_test.go:
```diff
@@ -41,10 +41,12 @@ import (
 	"github.com/onsi/gomega/ghttp"
 
 	k8sv1 "k8s.io/api/core/v1"
+	v12 "k8s.io/api/core/v1"
```
please use the already present `k8sv1`

fossedihelm commented on tests/storage/migration.go:
```diff
@@ -54,6 +54,7 @@ import (
 	"kubevirt.io/kubevirt/tests/framework/checks"
 	"kubevirt.io/kubevirt/tests/framework/kubevirt"
 	"kubevirt.io/kubevirt/tests/framework/matcher"
+	. "kubevirt.io/kubevirt/tests/framework/matcher"
```
duplicate import`

fossedihelm commented on tests/storage/migration.go:
```diff
@@ -539,6 +539,23 @@ var _ = SIGDescribe("[Serial]Volumes update with migration", Serial, func() {
+			restartOptions := &virtv1.RestartOptions{GracePeriodSeconds: pointer.P(int64(0))}
+			err := virtClient.VirtualMachine(vm.Namespace).Restart(context.Background(), vm.Name, restartOptions)
+			Expect(err).To(MatchError(ContainSubstring("VM recovery required")))
+		})
```
Can't this be merged with the other one for start?
I mean, first try to restart, ensure that the request is blocked, then stop and try to start

fossedihelm reviewed (commented):
Thank you! 
Some small comments below

alicefr commented:
I fixed the last comments and added also the unit tests for the `syncVolumeMigration`. In order to properly write the unit tests, I have added the `WithVMStatus` function.
@mhenriks @fossedihelm PTAL

@EdDev can you please take a look to commit 3a6460b58477dbec0964d8a2a6badc24bc9c9582 , it was required to write the status of the VM in the unit tests.

fossedihelm commented on pkg/libvmi/status/status.go:
```diff
@@ -133,3 +133,35 @@ func WithInterfaceStatus(interfaceStatus v1.VirtualMachineInstanceNetworkInterfa
+	return *vmStatus
+}
+
+func WithVolumeUpdateState(volumeUpdateState *v1.VolumeUpdateState) VMOption {
```
Currently it is not present but in future we could be add the `WithVolumeUpdateState` for vmi.
I would stay with using `VM` for VM related options
```suggestion
func WithVMVolumeUpdateState(volumeUpdateState *v1.VolumeUpdateState) VMOption {
```

mhenriks commented:
/lgtm
/hold

holding for @fossedihelm and @EdDev

fossedihelm commented:
/lgtm
Thank you @alicefr 
@EdDev up to you to unhold

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.30-sig-storage

alicefr commented:
/unhold
Got good feedback from @EdDev

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.30-sig-storage

fossedihelm commented:
/hold
@alicefr can you have a look at the failures?

alicefr commented:
/test pull-kubevirt-unit-test

fossedihelm commented:
/lgtm

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.30-sig-storage

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.30-sig-storage
